Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp

```diff
@@ -14,12 +14,186 @@
 #ifndef ECU_PE3_PE3_H
 #define ECU_PE3_PE3_H
 
+// Standard Libraries
+#include <inttypes.h>
+
+#include <vector>
+#include <array>
+#include <memory>
+
+// DFR Custom Dependencies
+#include "iecu.hpp"
+#include "../../../Platform/CAN/Interfaces/ican.hpp"
+
+// Internal ECU Messages
+#include "Frames/frame_pe1.hpp"
+#include "Frames/frame_pe2.hpp"
+#include "Frames/frame_pe3.hpp"
+#include "Frames/frame_pe4.hpp"
+#include "Frames/frame_pe5.hpp"
+#include "Frames/frame_pe6.hpp"
+#include "Frames/frame_pe7.hpp"
+#include "Frames/frame_pe8.hpp"
+#include "Frames/frame_pe9.hpp"
+#include "Frames/frame_pe10.hpp"
+#include "Frames/frame_pe11.hpp"
+#include "Frames/frame_pe12.hpp"
+#include "Frames/frame_pe13.hpp"
+#include "Frames/frame_pe14.hpp"
+#include "Frames/frame_pe15.hpp"
+#include "Frames/frame_pe16.hpp"
+
 namespace sensor {
 
-class Pe3 {
+#define FramePe1Id		0x0CFFF048
+#define FramePe2Id		0x0CFFF148
+#define FramePe3Id		0x0CFFF248
+#define FramePe4Id		0x0CFFF348
+#define FramePe5Id		0x0CFFF448
+#define FramePe6Id		0x0CFFF548
+#define FramePe7Id		0x0CFFF648
+#define FramePe8Id		0x0CFFF748
+#define FramePe9Id		0x0CFFF848
+#define FramePe10Id		0x0CFFF948
+#define FramePe11Id		0x0CFFFA48
+#define FramePe12Id		0x0CFFFB48
+#define FramePe13Id		0x0CFFFC48
+#define FramePe14Id		0x0CFFFD48
+#define FramePe15Id		0x0CFFFE48
+#define FramePe16Id		0x0CFFD048
+
+class Pe3 : IEcu {
+public:
+	Pe3(std::shared_ptr<platform::ICan> can_bus);
+
+	virtual ~Pe3();
+
+	const std::vector<uint32_t>& CanIdList() override;
+
+	uint32_t LatestCanId() override;
+
+	bool NewMessageArrived() override;
+
+	void Update() override;
+
+	// Frame PE1 Fields
+	uint16_t rpm_ = 0;
+	float tps_ = 0.0f;
+	float fuel_open_time_ = 0.0f;
+	float ignition_angle_ = 0.0f;
+
+	// Frame PE2 Fields
+	float barometer_pressure_ = 0.0f;
+	float map_ = 0.0f;
+	float lambda_ = 0.0f;
+	PressureType pressure_unit = PressureType::kUnknown;
+
+	// Frame PE3 and PE4 Fields
+	static constexpr uint8_t kNumAnalogPorts = 8;
+	std::array<float, kNumAnalogPorts> analog_inputs_;
+
+	//Frame PE5 Fields
+	static constexpr uint8_t kNumFrequencyPorts = 4;
+	std::array<float, kNumFrequencyPorts> frequencies_;
+
+	// Frame PE6 Fields
+	float battery_voltage_ = 0.0f;
+	float air_temp_ = 0.0f;
+	float coolant_temp_ = 0.0f;
+	TemperatureType temperature_unit_ = TemperatureType::kUnknown;
+
+	// Frame PE7 Fields
+	static constexpr uint8_t kNumAnalogThermistorPorts = 2;
+	std::array<float, kNumAnalogThermistorPorts> analog_inputs_thermistors_;
+
+	// Frame PE8 Fields
+	float rpm_rate_ = 0.0f;
+	float tps_rate_ = 0.0f;
+	float map_rate_ = 0.0f;
+	float maf_load_rate_ = 0.0f;
+
+	// Frame PE9 Fields
+	float measured_lambda_1_ = 0.0f;
+	float measured_lambda_2_ = 0.0f;
```

Comment:
I find it a little difficult justifying an array of just two elements (I would also initialize a constant for its size), but I have also considered using C++ [std::pair](https://www.geeksforgeeks.org/pair-in-cpp-stl/) since this is enumerated `1` and `2` on the datasheet.

Probably overkill — but I have no preference. I will leave it up to you.

Reply:
I'm honestly fine with whatever. Pick whatever makes the most convenience for you

Reply:
I just used `std::array` (not be confused with raw arrays) for their automatic bounds-checking at compile-time.